### ISSUE TITLE
cp2k 4.1 fails to build due to ICE

### ISSUE THREAD

**Issue by pawosm-arm** [flang-compiler/flang#178]
Tarball: cp2k-4.1.tar.bz2
Download page (sourceforge): https://sourceforge.net/projects/cp2k/files
Dependencies: libxc2 (libxc-2.2.3.tar.gz) (tested using statically built by flang with --enable-fortran and FCFLAGS="-fPIC -DPIC")
Bug:

```
error: forward reference and definition of global have different types
--
@build_deriv1_eri = linkonce global [6 x [6 x [6 x [6 x %struct.c_funptr.1]]]] undef
```

...on file: hfx_libint_wrapper.F (you can build it individually, when you build all required .mod files first, but see note below!)

Note: when compiling this file, -D__LIBINT must be passed to flang, otherwise only #else part of the code is built and it does not fail the compiler.

e.g.:
```
flang -D__FFTW3 -D__LIBINT -D__LIBXC2 -D__MPI_VERSION=3 -D__LIBINT_MAX_AM=7 -D__LIBDERIV_MAX_AM1=6 -D__MAX_CONTR=4 -D__parallel -D__SCALAPACK -O2 -ffast-math -Mfreeform -ffree-line-length-none -fopenmp -ftree-vectorize -funroll-loops -c hfx_libint_wrapper.F
```

**Comment by pawosm-arm** [comment]
I've edited IR file generated by flang2 so that definition is placed before reference and the error message became more obvious:

```
hfx_libint_wrapper-8506f9.ll:442:43: error: '@build_deriv1_eri' defined with type '[6 x [6 x [6 x [6 x %struct.c_funptr.1]]]]*'
        %76 = bitcast [1296 x %struct.c_funptr]* @build_deriv1_eri to i8*, !dbg !245
                                                 ^
```

**Comment by pawosm-arm** [comment]
Something minimal that reproduces this bug:

```
module somemod

  implicit none

  type :: some
    integer :: sth
  end type some

  type(some), dimension(0:4, 0:4, 0:4, 0:4), bind(c) :: sth

contains

  subroutine sub1

    implicit none

    type(some) :: sth1
 
    sth1 = sth(1, 2, 3, 4)
  end subroutine sub1
 
end module somemod
```

some observations:

- failing variable is the global module variable, it is required for having bind(c)
- failing variable is declared with bind(c), without it program compiles
- failing variable is of derived type, without it (e.g. when using integer type) program compiles
    Due to minimalism of this example, it's best to build it with -O0

**Comment by pawosm-arm** [comment]
A comment in llutil.c confirms just that:
```
      /* We chose to flatten Fortran array into single dimension array because
       * how the dinit processing was done and how we access to its address in
       * the ili, which is linearized.  Not really sure how it dwarf generation
       * should be done - wait until then ...
       */
```

**Comment by pawosm-arm** [comment]
A quick fix for this issue (and the uninitialized anum) can be as follows:

```
diff --git a/tools/flang2/flang2exe/llutil.c b/tools/flang2/flang2exe/llutil.c
index b437c59..9925ba0 100644
--- a/tools/flang2/flang2exe/llutil.c
+++ b/tools/flang2/flang2exe/llutil.c
@@ -3819,6 +3819,7 @@ get_ftn_extern_lltype(int sptr)
 LL_Type *
 get_ftn_cbind_lltype(int sptr)
 {
+#if 0
   DTYPE dtype = DTYPEG(sptr);
   int sdtype, tag, numdim, gblsym, d, iface, gs;
   ISZ_T anum;
@@ -3826,6 +3827,7 @@ get_ftn_cbind_lltype(int sptr)
   char *typed, *name;
   char tname[MXIDLN];
   ADSC *ad;
+#endif

   assert(CFUNCG(sptr), "Expected CBIND type", sptr, 4);

@@ -3839,6 +3841,7 @@ get_ftn_cbind_lltype(int sptr)
   if (SCG(sptr) == SC_STATIC) /* internal procedure bind(c) */
     return get_ftn_static_lltype(sptr);

+#if 0
   if (SCG(sptr) == SC_EXTERN) {
     sdtype = dtype;
     if (DTY(dtype) == TY_ARRAY)
@@ -3880,6 +3883,7 @@ get_ftn_cbind_lltype(int sptr)
       return llt;
     }
   }
+#endif
   return make_lltype_from_dtype(DTYPEG(sptr));
 }
```
Questions are:
- do we really need this disabled code?
- why only multi-dimensional arrays of derived type values are treated that way? (this code is never executed for arrays of built-in types!)

**Comment by gklimowicz** [comment]
It looks like this code was added as a workaround of some sort, but that the fundamental issue may have been fixed in another way. It's possible that this change is no longer needed. We're going to try the change you made and see if it causes any issue with the larger test suite that we run nightly.

**Comment by pawosm-arm** [comment]
Proposed change fixes this issue but causes test cases fs22406 and fs22408 to fail due to invalid IR generated ([99 x %struct.roo.1] instead of [99 x %struct.roo]).

**Comment by gklimowicz** [comment]
Should be fixed with commit bdb9f18.
